lib/semantics/check-do.cc

```diff
@@ -493,87 +434,137 @@ class DoContext {
-    // C1127
-    int count{0};
-    for (auto &ls : localitySpecs) {
-      if (std::holds_alternative<parser::LocalitySpec::DefaultNone>(ls.u)) {
-        ++count;
-        if (count > 1) {
-          context_.Say(currentStatementSourcePosition_,
-              "only one DEFAULT(NONE) may appear"_en_US);
-          return;
+    SymbolSet symbols;
+    const Scope &scope{
+        context_.FindScope(currentStatementSourcePosition_).parent()};
+    // Loop through the LocalitySpec::Local locality-specs
+    for (const auto &ls : localitySpecs) {
+      if (const auto *names{std::get_if<parser::LocalitySpec::Local>(&ls.u)}) {
+        // Loop through the names in the Local locality-spec getting their
+        // symbols
+        for (const parser::Name &name : names->v) {
+          if (const Symbol * symbol{scope.FindSymbol(name.source)}) {
+            if (const Symbol * root{GetAssociationRoot(*symbol)}) {
+              symbols.insert(root);
+            }
+          }
         }
       }
     }
+    return symbols;
   }
 
+  static SymbolSet GatherSymbolsFromExpression(const parser::Expr &expression) {
+    if (const auto *expr{GetExpr(expression)}) {
+      struct CollectSymbols : public virtual evaluate::VisitorBase<SymbolSet> {
+        explicit CollectSymbols(int) {}
+        void Handle(const Symbol *symbol) {
+          if (const Symbol * root{GetAssociationRoot(*symbol)}) {
+            result().insert(root);
+          }
+        }
+      };
+      return evaluate::Visitor<CollectSymbols>{0}.Traverse(*expr);
+    } else {
+      return {};
+    }
+  }
+
+  // C1121 - procedures in mask must be pure
   void CheckMaskIsPure(const parser::ScalarLogicalExpr &mask) const {
-    // C1121 - procedures in mask must be pure
-    // TODO - add the name of the impure procedure to the message
-    SymbolContainer references{
-        GatherReferencesFromExpression(mask.thing.thing.value())};
-    for (auto *r : references) {
-      if (isProcedure(r->flags()) && !isPure(r->attrs())) {
-        context_.Say(currentStatementSourcePosition_,
-            "concurrent-header mask expression cannot reference an impure"
-            " procedure"_err_en_US);
+    SymbolSet references{GatherSymbolsFromExpression(mask.thing.thing.value())};
+    for (const Symbol *ref : references) {
+      if (IsProcedure(*ref) && !IsPureProcedure(*ref)) {
+        const parser::CharBlock &name{ref->name()};
+        context_
+            .Say(currentStatementSourcePosition_,
+                "concurrent-header mask expression cannot reference an impure"
+                " procedure"_err_en_US)
+            .Attach(name, "Declaration of impure procedure '%s'"_en_US, name);
         return;
       }
     }
   }
 
-  void CheckNoCollisions(const SymbolContainer &refs,
-      const SymbolContainer &defs,
-      const parser::MessageFixedText &errorMessage) const {
+  void CheckNoCollisions(const SymbolSet &refs, const SymbolSet &set,
```

Comment:
why change name from `defs` to `set` ? 

`refs` and `defs` makes sense to me.

Reply:
The function ```CheckNoCollisions()``` is used to compare names in limit and step expressions against index names and to compare names in the ```scalar-mask-expr``` against names in ```locality-spec```s.  In the first case, the index-names can be considered to be definitions or defs.  But in the second case, there really aren't any definitions.  In both cases, though, the names in the expressions can be considered references or refs.

Reply:
Okay, then maybe you could name it `uses` or something that conveys what it is. `set` doesn't say anything really.

Reply:
Good suggestion.  I'll do that.  Thanks!